File: apps/livekit-worker/src/tools/leadTool.ts
import { llm } from '@livekit/agents';
import { z } from 'zod';
import type { RuntimeToolDefinition } from '@nextlite/shared';
import type { ToolRuntimeContext, ToolFactory, NativeLiveKitTool } from './toolRegistry.ts';

export const createCallbackLeadArgsSchema = z.object({
  customerName: z
    .string()
    .trim()
    .min(1, 'Customer name is required')
    .max(255, 'Customer name exceeds 255 characters')
    .describe('Full name of the customer requesting a callback or follow-up'),
  customerPhone: z
    .string()
    .trim()
    .max(50, 'Customer phone exceeds 50 characters')
    .optional()
    .describe(
      'Contact phone number for the callback. Omit or leave empty if the caller requests using their current/incoming phone number',
    ),
  customerEmail: z
    .string()
    .trim()
    .email('Invalid email address')
    .max(255)
    .optional()
    .or(z.literal(''))
    .describe('Optional email address of the customer'),
  interestCategory: z
    .string()
    .trim()
    .max(255)
    .optional()
    .describe(
      'Specific service, product, property, course, inquiry type, or subject the customer wants to discuss',
    ),
  notes: z
    .string()
    .trim()
    .max(2000, 'Notes exceed 2000 characters')
    .optional()
    .describe('Summary notes of the request, preferred callback time window, or key customer questions'),
  metadata: z
    .record(z.string(), z.unknown())
    .optional()
    .describe('Optional structured metadata attributes associated with the lead'),
});

export type CreateCallbackLeadArgs = z.infer<typeof createCallbackLeadArgsSchema>;

export interface LeadToolSuccessResult {
  success: true;
  leadId: string;
  message: string;
}

export interface LeadToolFailureResult {
  success: false;
  error: string;
  message: string;
}

export type LeadToolResult = LeadToolSuccessResult | LeadToolFailureResult;

export interface CreateCallbackLeadToolOptions {
  apiUrl?: string | undefined;
  workerSecret?: string | undefined;
  timeoutMs?: number | undefined;
  fetchFn?: typeof fetch | undefined;
}

/**
 * Creates the native LiveKit `create_callback_lead` function tool.
 * 
 * Security & Boundary Rules:
 * - `deploymentId` and `callSessionId` are injected exclusively from trusted `runtimeContext`.
 * - The LLM can never provide or override tenantId, agentId, or deploymentId.
 * - If `customerPhone` is omitted in the tool call, falls back to trusted `runtimeContext.callerPhone`.
 * - Structured success (`success: true`) or failure (`success: false`) is returned to prevent hallucinations.
 */
export function createCallbackLeadTool(
  context: ToolRuntimeContext,
  options?: CreateCallbackLeadToolOptions,
): NativeLiveKitTool {
  if (!context.deploymentId || context.deploymentId.trim() === '') {
    throw new Error('[LeadTool] deploymentId is required in runtime context to initialize create_callback_lead tool');
  }

  const trustedDeploymentId = context.deploymentId.trim();
  const trustedCallSessionId = context.callSessionId?.trim();
  const trustedCallerPhone = context.callerPhone?.trim();

  const baseUrl = (options?.apiUrl || context.apiUrl || process.env.NEXTLITE_API_URL || 'http://localhost:3001').replace(
    /\/+$/,
    '',
  );
  const workerSecret =
    options?.workerSecret || context.workerSecret || process.env.LIVEKIT_WORKER_SECRET || 'dev-livekit-worker-secret-v3';
  const timeoutMs = options?.timeoutMs || 4000;
  const customFetch = options?.fetchFn || context.fetchFn || globalThis.fetch;

  return llm.tool({
    name: 'create_callback_lead',
    description:
      'Record a customer lead or callback request when the caller explicitly asks for follow-up, requests a callback, or clearly wants a team member to contact them about a product or service.',
    parameters: createCallbackLeadArgsSchema,
    execute: async (args: CreateCallbackLeadArgs): Promise<LeadToolResult> => {
      // 1. Validate parameters with Zod schema
      const parsed = createCallbackLeadArgsSchema.safeParse(args);
      if (!parsed.success) {
        const errorMsg = parsed.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join('; ');
        return {
          success: false,
          error: 'INVALID_ARGUMENTS',
          message: `Invalid callback request details: ${errorMsg}`,
        };
      }

      const { customerName, customerPhone, customerEmail, interestCategory, notes, metadata } = parsed.data;

      // 2. Caller phone fallback: use explicit customerPhone if provided, otherwise fallback to trusted callerPhone
      let effectivePhone: string | undefined;
      if (customerPhone && customerPhone.trim().length > 0) {
        effectivePhone = customerPhone.trim();
      } else if (trustedCallerPhone && trustedCallerPhone.length > 0) {
        effectivePhone = trustedCallerPhone;
      }

      if (!effectivePhone) {
        return {
          success: false,
          error: 'INVALID_ARGUMENTS',
          message:
            'Customer phone number is required to record a callback request. Please ask the caller for their contact phone number.',
        };
      }

      // 3. Prepare payload for internal control-plane endpoint
      const requestPayload: Record<string, unknown> = {
        deploymentId: trustedDeploymentId,
        customerName: customerName.trim(),
        customerPhone: effectivePhone,
      };

      if (trustedCallSessionId) {
        requestPayload.callSessionId = trustedCallSessionId;
      }
      if (customerEmail && customerEmail.trim().length > 0) {
        requestPayload.customerEmail = customerEmail.trim();
      }
      if (interestCategory && interestCategory.trim().length > 0) {
        requestPayload.interestCategory = interestCategory.trim();
      }
      if (notes && notes.trim().length > 0) {
        requestPayload.notes = notes.trim();
      }
      if (metadata && typeof metadata === 'object' && Object.keys(metadata).length > 0) {
        requestPayload.metadata = metadata;
      }

      // 4. Send authenticated request to POST /api/internal/leads
      const url = `${baseUrl}/api/internal/leads`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), timeoutMs);

      try {
        const response = await customFetch(url, {
          method: 'POST',
          headers: {
            'x-worker-secret': workerSecret,
            'Authorization': `Bearer ${workerSecret}`,
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          body: JSON.stringify(requestPayload),
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (!response.ok) {
          let errorData: { error?: string; message?: string; code?: string } | null = null;
          try {
            errorData = (await response.json()) as { error?: string; message?: string; code?: string };
          } catch {
            errorData = null;
          }

          const status = response.status;
          const errorCode =
            errorData?.code ||
            (status === 401
              ? 'UNAUTHORIZED'
              : status === 404
                ? 'DEPLOYMENT_NOT_FOUND'
                : status === 409
                  ? 'DEPLOYMENT_INACTIVE'
                  : 'LEAD_CREATION_FAILED');

          console.warn(`[LeadTool] API returned status ${status} for lead creation:`, errorData?.error || errorData?.message);

          return {
            success: false,
            error: errorCode,
            message: 'Unable to record the callback request at this time. Please try again later.',
          };
        }

        const responseData = (await response.json()) as { id?: string };
        const leadId = responseData?.id || 'lead-created';

        return {
          success: true,
          leadId,
          message: 'Callback request recorded successfully. Our team will contact the customer.',
        };
      } catch (err: unknown) {
        clearTimeout(timeoutId);

        const isAbort = (err as Error)?.name === 'AbortError';
        const errorCode = isAbort ? 'SERVICE_UNAVAILABLE' : 'CONNECTION_ERROR';
        console.warn(`[LeadTool] Failed to connect to leads API:`, (err as Error)?.message || err);

        return {
          success: false,
          error: errorCode,
          message: 'Unable to record the callback request due to a temporary network issue. Please try again later.',
        };
      }
    },
  });
}

/**
 * Built-in tool factory for `create_callback_lead`.
 */
export const callbackLeadToolFactory: ToolFactory = {
  toolId: 'create_callback_lead',
  create(
    context: ToolRuntimeContext,
    toolConfig?: RuntimeToolDefinition,
  ): NativeLiveKitTool {
    const baseTool = createCallbackLeadTool(context, {
      apiUrl: context.apiUrl,
      workerSecret: context.workerSecret,
      fetchFn: context.fetchFn,
    });

    if (toolConfig?.description && toolConfig.description.trim()) {
      baseTool.description = toolConfig.description.trim();
    }

    return baseTool;
  },
};

File: apps/pipecat-worker/app/tools/lead_tool.py
"""NextLite Pipecat Create Callback Lead Tool.

Ports generic lead creation behavior from LiveKit worker to Pipecat native tool calling.
Connects strictly to NextLite Control Plane internal API (POST /api/internal/leads).

Security & Boundary Rules:
- deploymentId and callSessionId are injected exclusively from trusted ToolRuntimeContext.
- LLM cannot provide or override tenantId, agentId, deploymentId, or callSessionId.
- If customerPhone is omitted, falls back to trusted context.caller_phone.
- Never accesses PostgreSQL directly.
- Never generates lead IDs in Pipecat.
- Structured success or failure is returned to the LLM to prevent hallucinations.
"""

from typing import TYPE_CHECKING, Any, Callable, Dict, Optional
import httpx
from loguru import logger

from pipecat.adapters.schemas.function_schema import FunctionSchema
from pipecat.services.llm_service import FunctionCallParams

from app.turn_timing import log_phone_trace

if TYPE_CHECKING:
    from app.tools.tool_registry import ToolRuntimeContext

CREATE_CALLBACK_LEAD_TOOL_NAME = "create_callback_lead"

LEAD_TOOL_PROPERTIES: Dict[str, Any] = {
    "customerName": {
        "type": "string",
        "description": "Customer full name",
    },
    "customerPhone": {
        "type": "string",
        "description": "Contact phone if provided",
    },
    "notes": {
        "type": "string",
        "description": "Callback notes",
    },
}

LEAD_TOOL_REQUIRED = ["customerName"]


def create_callback_lead_tool_factory(
    context: "ToolRuntimeContext",
    description_override: Optional[str] = None,
    http_client: Optional[httpx.AsyncClient] = None,
) -> FunctionSchema:
    """Creates a native Pipecat FunctionSchema for create_callback_lead bound to trusted context."""
    if not context.deployment_id or not context.deployment_id.strip():
        raise ValueError("[LeadTool] deployment_id is required in ToolRuntimeContext")

    trusted_deployment_id = context.deployment_id.strip()

    base_url = (context.api_url or "http://localhost:3001").rstrip("/")
    worker_secret = context.worker_secret or "dev-worker-api-secret"
    api_endpoint = f"{base_url}/api/internal/leads"

    tool_description = (
        description_override.strip()
        if description_override and description_override.strip()
        else "Record a callback request or lead when the customer asks for follow-up."
    )

    async def handle_create_callback_lead(params: FunctionCallParams) -> Dict[str, Any]:
        await context.ensure_call_session_id()
        trusted_call_session_id = context.call_session_id.strip() if context.call_session_id else None
        trusted_caller_phone = context.caller_phone.strip() if context.caller_phone else None

        raw_args = params.arguments or {}
        if not isinstance(raw_args, dict):
            failure_result = {
                "success": False,
                "error": "INVALID_ARGUMENTS",
                "message": "Invalid callback request details: arguments must be a JSON object.",
            }
            if params.result_callback:
                await params.result_callback(failure_result)
            return failure_result

        # 1. Validate customerName
        raw_name = raw_args.get("customerName")
        if not raw_name or not isinstance(raw_name, str) or not raw_name.strip():
            failure_result = {
                "success": False,
                "error": "INVALID_ARGUMENTS",
                "message": "Customer name is required to record a callback request.",
            }
            if params.result_callback:
                await params.result_callback(failure_result)
            return failure_result

        customer_name = raw_name.strip()
        if len(customer_name) > 255:
            failure_result = {
                "success": False,
                "error": "INVALID_ARGUMENTS",
                "message": "Customer name exceeds 255 characters.",
            }
            if params.result_callback:
                await params.result_callback(failure_result)
            return failure_result

        # 2. Resolve caller phone: supplied customerPhone > trusted callerPhone fallback
        raw_phone = raw_args.get("customerPhone")
        turn_id = context.timing_tracker.active_turn_id if context.timing_tracker else "unknown"
        log_phone_trace("tool_argument_received", turn_id, raw_phone)

        effective_phone: Optional[str] = None
        if raw_phone and isinstance(raw_phone, str) and raw_phone.strip():
            effective_phone = raw_phone.strip()
        elif trusted_caller_phone:
            effective_phone = trusted_caller_phone

        log_phone_trace("control_plane_payload", turn_id, effective_phone)

        if not effective_phone:
            failure_result = {
                "success": False,
                "error": "INVALID_ARGUMENTS",
                "message": (
                    "Customer phone number is required to record a callback request. "
                    "Please ask the caller for their contact phone number."
                ),
            }
            if params.result_callback:
                await params.result_callback(failure_result)
            return failure_result

        # 3. Assemble API payload strictly with trusted identity fields
        payload: Dict[str, Any] = {
            "deploymentId": trusted_deployment_id,
            "customerName": customer_name,
            "customerPhone": effective_phone,
        }
        if trusted_call_session_id:
            payload["callSessionId"] = trusted_call_session_id

        raw_email = raw_args.get("customerEmail")
        if raw_email and isinstance(raw_email, str) and raw_email.strip():
            payload["customerEmail"] = raw_email.strip()

        raw_category = raw_args.get("interestCategory")
        if raw_category and isinstance(raw_category, str) and raw_category.strip():
            payload["interestCategory"] = raw_category.strip()

        raw_notes = raw_args.get("notes")
        if raw_notes and isinstance(raw_notes, str) and raw_notes.strip():
            payload["notes"] = raw_notes.strip()

        raw_meta = raw_args.get("metadata")
        if raw_meta and isinstance(raw_meta, dict) and len(raw_meta) > 0:
            payload["metadata"] = raw_meta

        headers = {
            "x-worker-secret": worker_secret,
            "Authorization": f"Bearer {worker_secret}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

        # 4. Perform authenticated HTTP call to NextLite Control Plane
        result: Dict[str, Any]
        try:
            if http_client:
                response = await http_client.post(
                    api_endpoint,
                    json=payload,
                    headers=headers,
                    timeout=4.0,
                )
            else:
                async with httpx.AsyncClient(timeout=4.0) as client:
                    response = await client.post(
                        api_endpoint,
                        json=payload,
                        headers=headers,
                    )

            if response.status_code in (200, 201):
                resp_json = response.json() if response.content else {}
                lead_id = resp_json.get("id", "lead-created")
                result = {
                    "success": True,
                    "leadId": lead_id,
                    "message": "Callback request recorded successfully. Our team will contact the customer.",
                }
            else:
                err_data: Dict[str, Any] = {}
                try:
                    err_data = response.json()
                except Exception:
                    pass

                status = response.status_code
                error_code = err_data.get("code") or (
                    "UNAUTHORIZED" if status == 401
                    else "DEPLOYMENT_NOT_FOUND" if status == 404
                    else "DEPLOYMENT_INACTIVE" if status == 409
                    else "LEAD_CREATION_FAILED"
                )
                logger.warning(
                    f"[LeadTool] API returned status {status} for lead creation: {err_data.get('error') or err_data.get('message')}"
                )
                result = {
                    "success": False,
                    "error": error_code,
                    "message": "Unable to record the callback request at this time. Please try again later.",
                }
        except httpx.TimeoutException:
            logger.warning("[LeadTool] Timeout connecting to leads API")
            result = {
                "success": False,
                "error": "SERVICE_UNAVAILABLE",
                "message": "Unable to record the callback request due to a temporary network timeout. Please try again later.",
            }
        except Exception as err:
            logger.warning(f"[LeadTool] Failed to connect to leads API: {err}")
            result = {
                "success": False,
                "error": "CONNECTION_ERROR",
                "message": "Unable to record the callback request due to a temporary network issue. Please try again later.",
            }

        # 5. Deliver result to Pipecat LLM callback
        if params.result_callback:
            await params.result_callback(result)

        return result

    return FunctionSchema(
        name=CREATE_CALLBACK_LEAD_TOOL_NAME,
        description=tool_description,
        properties=LEAD_TOOL_PROPERTIES,
        required=LEAD_TOOL_REQUIRED,
        handler=handle_create_callback_lead,
    )
